### PR TITLE
PLATFORM-1850 fixing bad iteration

### PR DESCRIPTION
**Pull request by pgrabowski-wikia**
@macbre @Grunny 

https://wikia-inc.atlassian.net/browse/PLATFORM-1850

### CODE REVIEW

File: extensions/MultiUpload/multiupload.js [hunk 1]
```diff
@@ -266,10 +266,10 @@ function fillDestFilename( id ) {
...
 	if( wgFileExtensions && id != 'wpUploadFileURL' ) {
 		var found = false;
-		if( fname.lastIndexOf( '.' ) != -1 ) {
-			var ext = fname.substr( fname.lastIndexOf( '.' ) + 1 );
-			for( var i = 0; i < wgFileExtensions.length; i++ ) {
-				if( wgFileExtensions[i].toLowerCase() == ext.toLowerCase() ) {
+		if ( fname.lastIndexOf('.') != -1 ) {
+			var ext = fname.substr( fname.lastIndexOf('.') + 1 ).toLowerCase();
+			for (var wgfi in wgFileExtensions) {
+				if ( wgFileExtensions[wgfi].toLowerCase() == ext ) {
```
Comment (WikiaTech) [inline extensions/MultiUpload/multiupload.js]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANestedIfDepth)
MAJOR Replace "==" with "===". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)

File: extensions/MultiUpload/multiupload.js [hunk 2]
```diff
@@ -266,10 +266,10 @@ function fillDestFilename( id ) {
 	current = id.replace( /[^0-9]/g, '' );
 	if( wgFileExtensions && id != 'wpUploadFileURL' ) {
```
Comment (WikiaTech) [inline extensions/MultiUpload/multiupload.js]: MAJOR Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)

File: extensions/MultiUpload/multiupload.js [hunk 3]
```diff
@@ -266,10 +266,10 @@ function fillDestFilename( id ) {
...
 		var found = false;
-		if( fname.lastIndexOf( '.' ) != -1 ) {
-			var ext = fname.substr( fname.lastIndexOf( '.' ) + 1 );
-			for( var i = 0; i < wgFileExtensions.length; i++ ) {
-				if( wgFileExtensions[i].toLowerCase() == ext.toLowerCase() ) {
+		if ( fname.lastIndexOf('.') != -1 ) {
```
Comment (WikiaTech) [inline extensions/MultiUpload/multiupload.js]: MAJOR Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 54 issues
* CRITICAL 1 critical
* MAJOR 53 major

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [multiupload.js#L360](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L360): Remove this use of the "eval" function. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEval)
1. MAJOR [multiupload.js#L6](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L6): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
1. MAJOR [multiupload.js#L7](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L7): Replace "==" with "===". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L13](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L13): Move the declaration of "wgUploadLicenseObj" before this usage. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)
1. MAJOR [multiupload.js#L16](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L16): Function has a complexity of 14 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. MAJOR [multiupload.js#L19](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L19): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
1. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L36](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L36): Remove or correct this useless self-assignment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1656)
1. MAJOR [multiupload.js#L43](https://github.com/Wikia/app/blob/b5751c0aeb7807dbfc892219ef7bc97adb0edd72/extensions/MultiUpload/multiupload.js#L43): Rename "i" as this name is already used in declaration at line 35. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS2814)

File: extensions/MultiUpload/multiupload.js [hunk 4]
```diff
@@ -266,10 +266,10 @@ function fillDestFilename( id ) {
...
 	if( wgFileExtensions && id != 'wpUploadFileURL' ) {
 		var found = false;
-		if( fname.lastIndexOf( '.' ) != -1 ) {
-			var ext = fname.substr( fname.lastIndexOf( '.' ) + 1 );
-			for( var i = 0; i < wgFileExtensions.length; i++ ) {
-				if( wgFileExtensions[i].toLowerCase() == ext.toLowerCase() ) {
+		if ( fname.lastIndexOf('.') !== -1 ) {
+			var ext = fname.substr( fname.lastIndexOf('.') + 1 ).toLowerCase();
+			for (var wgfi in wgFileExtensions) {
+				if ( wgFileExtensions[wgfi].toLowerCase() === ext ) {
```
Comment (WikiaTech) [inline extensions/MultiUpload/multiupload.js:272]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ANestedIfDepth)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 52 issues
* CRITICAL 1 critical
* MAJOR 51 major

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [multiupload.js#L360](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L360): Remove this use of the "eval" function. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEval)
1. MAJOR [multiupload.js#L6](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L6): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
1. MAJOR [multiupload.js#L7](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L7): Replace "==" with "===". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L13](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L13): Move the declaration of "wgUploadLicenseObj" before this usage. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)
1. MAJOR [multiupload.js#L16](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L16): Function has a complexity of 14 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
1. MAJOR [multiupload.js#L19](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L19): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
1. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
1. MAJOR [multiupload.js#L36](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L36): Remove or correct this useless self-assignment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1656)
1. MAJOR [multiupload.js#L43](https://github.com/Wikia/app/blob/c58b5e6d330af60e9797b396f71c146c83600bf7/extensions/MultiUpload/multiupload.js#L43): Rename "i" as this name is already used in declaration at line 35. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS2814)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 51 issues
- CRITICAL 1 critical
- MAJOR 50 major

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. CRITICAL [multiupload.js#L360](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L360): Remove this use of the "eval" function. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEval)
2. MAJOR [multiupload.js#L6](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L6): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
3. MAJOR [multiupload.js#L7](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L7): Replace "==" with "===". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
4. MAJOR [multiupload.js#L13](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L13): Move the declaration of "wgUploadLicenseObj" before this usage. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AVariableDeclarationAfterUsage)
5. MAJOR [multiupload.js#L16](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L16): Function has a complexity of 14 which is greater than 10 authorized. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AFunctionComplexity)
6. MAJOR [multiupload.js#L19](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L19): Merge this if statement with the nested one. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3ACollapsibleIfStatements)
7. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
8. MAJOR [multiupload.js#L33](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L33): Replace "!=" with "!==". [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AEqEqEq)
9. MAJOR [multiupload.js#L36](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L36): Remove or correct this useless self-assignment. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS1656)
10. MAJOR [multiupload.js#L43](https://github.com/Wikia/app/blob/00e68764358e888f4bdae222ebb4e2eee854e310/extensions/MultiUpload/multiupload.js#L43): Rename "i" as this name is already used in declaration at line 35. [rule](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AS2814)

File: extensions/MultiUpload/multiupload.js [hunk 5]
```diff
@@ -264,12 +264,12 @@ function fillDestFilename( id ) {
...
-	if( wgFileExtensions && id != 'wpUploadFileURL' ) {
+	if( wgFileExtensions && id !== 'wpUploadFileURL' ) {
 		var found = false;
-		if( fname.lastIndexOf( '.' ) != -1 ) {
-			var ext = fname.substr( fname.lastIndexOf( '.' ) + 1 );
-			for( var i = 0; i < wgFileExtensions.length; i++ ) {
-				if( wgFileExtensions[i].toLowerCase() == ext.toLowerCase() ) {
+		if ( fname.lastIndexOf('.') !== -1 ) {
+			var ext = fname.substr( fname.lastIndexOf('.') + 1 ).toLowerCase();
+			for (var wgfi in wgFileExtensions) {
```
Comment (macbre) [inline extensions/MultiUpload/multiupload.js:271]: Why are we changing from iteration over array items to `foo in bar`?

Comment (pgrabowski-wikia) [reply extensions/MultiUpload/multiupload.js:271]: wgFileExtensions is an object, not an array.

Comment (macbre) [reply extensions/MultiUpload/multiupload.js:271]: [The upstream version of MultiUpload](https://github.com/wikimedia/mediawiki-extensions-MultiUpload/blob/bfbfe7c6787b9201704e7abe6807031bc8cd076d/resources/upload.js.patched#L391) assumes it's an array. Interesting, did we change the way `wgFileExtensions` is exposed to JavaScript?

``` php
includes/DefaultSettings.php
526:$wgFileExtensions = array( 'png', 'gif', 'jpg', 'jpeg' );

extensions/MultiUpload/MultiUpload.body.php
541:        global $wgExtensionsPath, $wgFileExtensions;
565:            'wgFileExtensions' => $wgFileExtensions
```

Or maybe `wgFileExtensions` value is customized via WikiFactory on a specific wiki and that's the root cause.

Comment (Grunny) [reply extensions/MultiUpload/multiupload.js:271]: @macbre It was an issue with the saved WF value for `wgFileExtensionsLocal` on Tardis. I fixed it there and it's now an array as expected again. :).

Comment (macbre) [reply extensions/MultiUpload/multiupload.js:271]: @Grunny - jolly good :)

**Pull request closed without merging**